File: vivordo_health/lib/src/utils/activity_score.dart
class ActivityScoreResult {
  const ActivityScoreResult({
    required this.score,
    required this.availableSignals,
  });

  final double score;
  final int availableSignals;
}

/// Scores daily activity from movement, intentional exercise, and exertion.
///
/// Steps and exercise are interchangeable ways to demonstrate activity: the
/// stronger of those two signals receives the primary weight and the other
/// receives a smaller complementary weight. Active calories capture exertion.
/// Missing signals are excluded and the remaining weights are normalized;
/// recorded zeroes remain part of the calculation.
ActivityScoreResult? calculateActivityScore({
  required double? steps,
  required double? exerciseMinutes,
  required double? activeCalories,
  required double stepsGoal,
  required double exerciseMinutesGoal,
  required double activeCaloriesGoal,
}) {
  double? normalized(double? value, double goal) {
    if (value == null || !value.isFinite || goal <= 0 || !goal.isFinite) {
      return null;
    }
    return (value / goal * 100).clamp(0.0, 100.0).toDouble();
  }

  final movement = normalized(steps, stepsGoal);
  final exercise = normalized(exerciseMinutes, exerciseMinutesGoal);
  final exertion = normalized(activeCalories, activeCaloriesGoal);
  final movementSignals = [?movement, ?exercise]
    ..sort((a, b) => b.compareTo(a));

  var weightedScore = 0.0;
  var availableWeight = 0.0;
  if (movementSignals.isNotEmpty) {
    weightedScore += movementSignals.first * 0.60;
    availableWeight += 0.60;
  }
  if (movementSignals.length > 1) {
    weightedScore += movementSignals[1] * 0.20;
    availableWeight += 0.20;
  }
  if (exertion != null) {
    weightedScore += exertion * 0.20;
    availableWeight += 0.20;
  }

  if (availableWeight == 0) return null;
  return ActivityScoreResult(
    score: (weightedScore / availableWeight).clamp(0.0, 100.0),
    availableSignals:
        (movement == null ? 0 : 1) +
        (exercise == null ? 0 : 1) +
        (exertion == null ? 0 : 1),
  );
}

File: vivordo_health/functions/activity_score.js
"use strict";

const DEFAULT_ACTIVITY_GOALS = Object.freeze({
  steps: 10000,
  exerciseMinutes: 40,
  activeCalories: 700,
});

/**
 * Reads validated activity goals from a Firestore user document.
 *
 * @param {Object|undefined} userData Firestore user data.
 * @return {{steps: number, exerciseMinutes: number, activeCalories: number}}
 */
function activityGoalsFromUserData(userData) {
  const stored = userData?.preferences?.activityGoals || {};
  const positive = (value, fallback) =>
    Number.isFinite(Number(value)) && Number(value) > 0 ?
      Number(value) : fallback;
  return {
    steps: positive(stored.steps, DEFAULT_ACTIVITY_GOALS.steps),
    exerciseMinutes: positive(
        stored.exerciseMinutes,
        DEFAULT_ACTIVITY_GOALS.exerciseMinutes,
    ),
    activeCalories: positive(
        stored.activeCalories,
        DEFAULT_ACTIVITY_GOALS.activeCalories,
    ),
  };
}

/**
 * Calculates activity while excluding unavailable signals and retaining zeroes.
 *
 * @param {Object} values Activity values and goals.
 * @return {{score: number, availableSignals: number}|null} Score result.
 */
function calculateActivityScore(values) {
  const normalized = (value, goal) => {
    if (!Number.isFinite(value) || !Number.isFinite(goal) || goal <= 0) {
      return null;
    }
    return Math.max(0, Math.min(100, value / goal * 100));
  };
  const movement = normalized(values.steps, values.stepsGoal);
  const exercise = normalized(
      values.exerciseMinutes,
      values.exerciseMinutesGoal,
  );
  const exertion = normalized(
      values.activeCalories,
      values.activeCaloriesGoal,
  );
  const movementSignals = [movement, exercise]
      .filter((value) => value !== null)
      .sort((a, b) => b - a);

  let weightedScore = 0;
  let availableWeight = 0;
  if (movementSignals.length > 0) {
    weightedScore += movementSignals[0] * 0.60;
    availableWeight += 0.60;
  }
  if (movementSignals.length > 1) {
    weightedScore += movementSignals[1] * 0.20;
    availableWeight += 0.20;
  }
  if (exertion !== null) {
    weightedScore += exertion * 0.20;
    availableWeight += 0.20;
  }
  if (availableWeight === 0) return null;

  return {
    score: Math.max(0, Math.min(100, weightedScore / availableWeight)),
    availableSignals: [movement, exercise, exertion]
        .filter((value) => value !== null).length,
  };
}

module.exports = {
  activityGoalsFromUserData,
  calculateActivityScore,
  DEFAULT_ACTIVITY_GOALS,
};
